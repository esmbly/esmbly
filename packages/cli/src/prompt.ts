import inquirer from 'inquirer';
import { Config } from '@esmbly/types';
import {
  getAvailableOutputFormats,
  getAvailableTransformers,
} from '@esmbly/utils';

export async function promptForConfig(): Promise<Config> {
  const { input } = await inquirer.prompt([
    {
      message: 'Which files do you want to transform?',
      name: 'input',
      type: 'input',
    },
  ]);

  const { transformers } = await inquirer.prompt([
    {
      choices: await getAvailableTransformers(),
      message: 'Which transformer do you want to use?',
      name: 'transformers',
      type: 'checkbox',
    },
  ]);

  // TODO: Ask to install transformers here?

  const { output } = await inquirer.prompt([
    {
      choices: await getAvailableOutputFormats(transformers),
      message: 'Which output formats to you want to use?',
      name: 'output',
      type: 'checkbox',
    },
  ]);

  return { input: [input], output, transformers };
}
