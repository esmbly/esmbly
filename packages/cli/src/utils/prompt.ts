import inquirer from 'inquirer';
import {
  TransformerInfo,
  getAvailableOutputFormats,
  getAvailableTransformers,
} from '@esmbly/utils';
import stringify from 'stringify-object';
import chalk from 'chalk';

interface PromptedConfig {
  module: string;
  install: boolean;
  transformers: TransformerInfo[];
  pkgManager: string;
}

function createRequireString(transformers: TransformerInfo[]): string {
  return transformers
    .map((t: TransformerInfo) => `const ${t.const} = require('${t.pkg}')`)
    .join('\n');
}

function createConfigString(
  input: string,
  output: string[],
  outDir: string,
): string {
  return stringify({
    input: [input],
    transformers: '<TRANSFORMERS>',
    // eslint-disable-next-line
    output: output.map((format: string) => ({ format, outDir })),
  });
}

function createTransformerCalls(transformers: TransformerInfo[]): string {
  /**
   * Produces:
   * [
   *   Flow.createTransformer(),
   *   Wasm.createTransformer()
   * ]
   */
  return `${transformers.reduce(
    (calls, t) => `${calls}\n    ${t.const}.createTransformer(),`,
    '[',
  )}\n  ]`;
}

function createModuleString(
  input: string,
  transformers: TransformerInfo[],
  output: string[],
  outDir: string,
): string {
  const requireString = createRequireString(transformers);
  const configString = createConfigString(input, output, outDir);
  const transformerCalls = createTransformerCalls(transformers);

  return `${requireString}\n\nmodule.exports = ${configString.replace(
    `'<TRANSFORMERS>'`,
    transformerCalls,
  )}\n`;
}

function isValidCombination(transformers: TransformerInfo[]): boolean {
  let isValid = true;

  for (let i = 0; i < transformers.length - 1; i += 1) {
    const current = transformers[i];
    const next = transformers[i + 1];

    if (!current.to.includes(next.from)) {
      isValid = false;
      break;
    }
  }

  return isValid;
}

export async function promptForConfig(): Promise<PromptedConfig> {
  const availableTransformers = getAvailableTransformers();

  const { input } = await inquirer.prompt([
    {
      message: 'Which files do you want to transform?',
      name: 'input',
      suffix: chalk.dim(' e.g ./**/*.js'),
      type: 'input',
    },
  ]);

  const { selectedTransformers } = await inquirer.prompt([
    {
      choices: availableTransformers.map(t => ({
        name: `${t.name}\t${chalk.dim(`${t.from} -> ${t.to.join(' | ')}`)}`,
        short: t.name,
        value: t.name,
      })),
      default: [],
      message: 'Which transformers do you want to use?',
      name: 'selectedTransformers',
      type: 'checkbox',
    },
  ]);

  const transformers = availableTransformers.filter(t =>
    selectedTransformers.includes(t.name),
  );

  if (!isValidCombination(transformers)) {
    const combination = transformers.reduce(
      (str: string, transformer: TransformerInfo, i: number) => {
        if (i === 0) {
          return `${transformer.name}`;
        }

        return `${str} -> ${transformer.name}`;
      },
      '',
    );
    throw new Error(
      `Transformations in the order ${combination} is not possible`,
    );
  }

  const { output } = await inquirer.prompt([
    {
      choices: getAvailableOutputFormats(transformers),
      message: 'Which formats to you want to output?',
      name: 'output',
      type: 'checkbox',
      when: transformers.length > 0,
    },
  ]);

  const { outDir } = await inquirer.prompt([
    {
      message: 'To which directory do you want to output files?',
      name: 'outDir',
      suffix: chalk.dim(' e.g ./dist'),
      type: 'input',
    },
  ]);

  const { install } = await inquirer.prompt([
    {
      default: true,
      message: 'Do you want to install the selected transformers and the CLI?',
      name: 'install',
      type: 'confirm',
      when: transformers.length > 0,
    },
  ]);

  const { pkgManager } = await inquirer.prompt([
    {
      choices: ['npm', 'yarn'],
      message: 'Which package manager do you want to use for installation?',
      name: 'pkgManager',
      type: 'list',
      when: install === true,
    },
  ]);

  return {
    install,
    module: createModuleString(input, transformers, output, outDir),
    pkgManager,
    transformers,
  };
}
