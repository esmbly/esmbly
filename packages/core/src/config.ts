import { RunConfig, Transformer, OutputFormat, Output } from '@esmbly/types';
import * as errors from './errors';

export function validateRunConfig(config: RunConfig): void {
  if (!config || !config.input || !config.transformers || !config.output) {
    throw new Error(errors.MissingConfig());
  }
  if (config.input.length < 1) {
    throw new Error(errors.NoInput());
  }
  if (config.transformers.length < 1) {
    throw new Error(errors.NoTransformers());
  }
  if (config.output.length < 1) {
    throw new Error(errors.NoOutput());
  }
  config.transformers.forEach((transformer: Transformer) => {
    if (typeof transformer.run !== 'function') {
      throw new Error(errors.InvalidTransformer(transformer));
    }
  });
  config.output.forEach((out: Output) => {
    let format = typeof out === 'string' ? out : out.format;
    if (format in OutputFormat === false) {
      throw new Error(errors.InvalidOutput(format));
    }
  });
}
