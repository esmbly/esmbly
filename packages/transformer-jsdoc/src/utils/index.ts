export * from './toKeyword';
export * from './toTypeAnnotation';
export * from './isParam';
export * from './isReturn';
