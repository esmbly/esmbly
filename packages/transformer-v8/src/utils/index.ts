export * from './toKeyword';
export * from './toTypeAnnotation';
export * from './Profiler';
