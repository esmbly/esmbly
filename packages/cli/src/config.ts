import path from 'path';
import stringify from 'stringify-object';
import { Config, InitOptions } from '@esmbly/types';
import {
  exists,
  getRelativePathTo,
  getRoot,
  readFile,
  writeFile,
} from '@esmbly/utils';
import { promptForConfig } from './prompt';

export const DEFAULT_FILE = '.esmblyrc.js';

export async function getTemplateConfig(): Promise<string> {
  const templatePath = path.resolve(__dirname, `../defaults/${DEFAULT_FILE}`);
  const template = await readFile(templatePath);
  return template.toString();
}

export function getDefaultConfigPath(): string {
  const root = getRoot();
  return path.join(root, DEFAULT_FILE);
}

export async function readConfig(customPath?: string): Promise<Config[]> {
  const configPath = customPath || (await getDefaultConfigPath());
  const config = require(configPath) // eslint-disable-line
  const toArray = (maybeArrayConfig: Config | Config[]): Config[] => {
    if (Array.isArray(maybeArrayConfig)) {
      return maybeArrayConfig;
    }
    return [maybeArrayConfig];
  };
  if (typeof config === 'function') {
    const configFnResult = await config();
    return toArray(configFnResult);
  }
  return toArray(config);
}

export async function createConfig(
  options: InitOptions,
): Promise<{
  root: string;
  fileName: string;
}> {
  const root = getRoot();
  const defaultConfigPath = getDefaultConfigPath();
  if ((await exists(defaultConfigPath)) && !options.force) {
    const relativeConfigPath = getRelativePathTo(defaultConfigPath);
    throw new Error(
      `Config ${relativeConfigPath} already exists. Remove that file first.`,
    );
  } else if (options.default) {
    const template = await getTemplateConfig();
    await writeFile(defaultConfigPath, template, { overwrite: options.force });
  } else {
    const config = await promptForConfig();
    const content = `module.exports = ${stringify(config)}\n`;
    await writeFile(defaultConfigPath, content, { overwrite: options.force });
  }
  return { fileName: DEFAULT_FILE, root };
}
