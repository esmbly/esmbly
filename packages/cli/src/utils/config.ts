import path from 'path';
import { Config, InitOptions } from '@esmbly/types';
import {
  exists,
  getRelativePathTo,
  getRoot,
  readFile,
  writeFile,
} from '@esmbly/utils';
import { promptForConfig } from './prompt';
import { installCLI, installTransformers } from './install';

export const DEFAULT_FILE = 'esmbly.config.js';

export async function getTemplateConfig(): Promise<string> {
  const templatePath = path.resolve(
    __dirname,
    `../../defaults/${DEFAULT_FILE}`,
  );
  const template = await readFile(templatePath);
  return template.toString();
}

export function getDefaultConfigPath(): string {
  const root = getRoot();
  return path.join(root, DEFAULT_FILE);
}

type ConfigFnA = () => Config;
type ConfigFnB = () => Config[];
type ConfigFn = ConfigFnA | ConfigFnB;

function toArrayConfig(maybeArrayConfig: Config | Config[]): Config[] {
  if (Array.isArray(maybeArrayConfig)) {
    return maybeArrayConfig;
  }

  return [maybeArrayConfig];
}

export async function readConfig(
  customPath?: string,
  requirer: (requirePath: string) => unknown = require,
): Promise<Config[]> {
  const configPath = customPath || (await getDefaultConfigPath());
  const config = requirer(configPath) as Config | Config[] | ConfigFn;

  if (typeof config === 'function') {
    const configFnResult = await config();
    return toArrayConfig(configFnResult);
  }

  return toArrayConfig(config);
}

export async function createConfig(
  options: InitOptions,
): Promise<{
  root: string;
  fileName: string;
}> {
  const root = getRoot();
  const defaultConfigPath = getDefaultConfigPath();

  if ((await exists(defaultConfigPath)) && !options.force) {
    const relativeConfigPath = getRelativePathTo(defaultConfigPath);
    throw new Error(
      `Config ${relativeConfigPath} already exists. Remove that file first.`,
    );
  } else if (options.default) {
    const template = await getTemplateConfig();
    await writeFile(defaultConfigPath, template, { overwrite: options.force });
  } else {
    const config = await promptForConfig();

    if (config.install) {
      await installTransformers(config.transformers, config.pkgManager);
      await installCLI(config.pkgManager);
    }

    await writeFile(defaultConfigPath, config.module, {
      overwrite: options.force,
    });
  }

  return { fileName: DEFAULT_FILE, root };
}
