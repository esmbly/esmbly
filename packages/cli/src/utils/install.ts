import { exec } from 'child_process';
import { promisify } from 'util';
import { TransformerInfo } from '@esmbly/utils';
import ora from 'ora';

export function getPkgManagerCommand(pkgManager: string): string {
  switch (pkgManager) {
    case 'npm':
      return 'npm install --save-dev';
    case 'yarn':
      return 'yarn add --dev';
    default:
      throw new Error(`Package manager ${pkgManager} is not supported`);
  }
}

export async function installTransformers(
  transformers: TransformerInfo[],
  pkgManager: string,
): Promise<void> {
  const command = getPkgManagerCommand(pkgManager);

  for (const transformer of transformers) {
    const spinner = ora(`Installing: ${transformer.pkg}`).start();
    await promisify(exec)(`${command} ${transformer.pkg}`);
    spinner.succeed(`Installed: ${transformer.pkg}`);
  }
}

export async function installCLI(pkgManager: string): Promise<void> {
  const command = getPkgManagerCommand(pkgManager);

  const spinner = ora('Installing: @esmbly/cli').start();
  await promisify(exec)(`${command} @esmbly/cli`);
  spinner.succeed('Installed: @esmbly/core');
}
