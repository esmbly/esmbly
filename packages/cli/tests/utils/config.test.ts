import * as utils from '@esmbly/utils';
import * as prompt from '../../src/utils/prompt';
import {
  createConfig,
  getDefaultConfigPath,
  getTemplateConfig,
  readConfig,
} from '../../src/utils/config';
import * as MockConfig from '../__fixtures__/config';

function setup(
  shouldExist: boolean,
): {
  existsSpy: jest.SpyInstance;
  getRootSpy: jest.SpyInstance;
  promptSpy: jest.SpyInstance;
  writeFileSpy: jest.SpyInstance;
  tearDown: () => void;
} {
  const getRootSpy = jest.spyOn(utils, 'getRoot');
  const promptSpy = jest.spyOn(prompt, 'promptForConfig');
  const writeFileSpy = jest.spyOn(utils, 'writeFile');
  const existsSpy = jest.spyOn(utils, 'exists');
  getRootSpy.mockReturnValue('root-path');
  promptSpy.mockResolvedValue({
    install: false,
    module: 'config-module-source',
    pkgManager: 'npm',
    transformers: [],
  });
  writeFileSpy.mockResolvedValue();
  existsSpy.mockResolvedValue(shouldExist);
  return {
    existsSpy,
    getRootSpy,
    promptSpy,
    tearDown: (): void => {
      existsSpy.mockRestore();
      getRootSpy.mockRestore();
      promptSpy.mockRestore();
      writeFileSpy.mockRestore();
    },
    writeFileSpy,
  };
}

describe('createConfig', () => {
  it('creates a new config file', async () => {
    const { promptSpy, writeFileSpy, tearDown } = setup(false);
    const config = await createConfig({ default: false, force: false });
    expect(promptSpy).toHaveBeenCalledTimes(1);
    expect(writeFileSpy.mock.calls).toMatchSnapshot();
    expect(config).toEqual({ fileName: 'esmbly.config.js', root: 'root-path' });
    tearDown();
  });

  it('creates a new config file with default values when passing the default option', async () => {
    const { promptSpy, writeFileSpy, tearDown } = setup(false);
    await createConfig({ default: true, force: false });
    expect(promptSpy).not.toHaveBeenCalled();
    expect(writeFileSpy.mock.calls).toMatchSnapshot();
    tearDown();
  });

  it('throws an error if there already is a config file', async () => {
    const { promptSpy, tearDown } = setup(true);
    await expect(
      createConfig({ default: false, force: false }),
    ).rejects.toThrowErrorMatchingSnapshot();
    expect(promptSpy).not.toHaveBeenCalled();
    tearDown();
  });

  it('overwrites any existing config file when passing the force option', async () => {
    const { promptSpy, tearDown } = setup(true);
    const config = await createConfig({ default: false, force: true });
    expect(promptSpy).toHaveBeenCalled();
    expect(config).toEqual({ fileName: 'esmbly.config.js', root: 'root-path' });
    tearDown();
  });
});

describe('getDefaultConfigPath', () => {
  it('returns the correct default config path', () => {
    const getRootSpy = jest.spyOn(utils, 'getRoot');
    getRootSpy.mockReturnValue('root-path');
    expect(getDefaultConfigPath()).toEqual('root-path/esmbly.config.js');
  });
});

describe('getTemplateConfig', () => {
  it('resolves to the correct template config', async () => {
    await expect(getTemplateConfig()).toMatchSnapshot();
  });
});

describe('readConfig', () => {
  it('requires the config file from the default path', async () => {
    const requirer = jest.fn();
    requirer.mockReturnValue(MockConfig.config);
    const config = await readConfig(undefined, requirer);
    expect(config).toEqual(MockConfig.config);
  });

  it('requires the config file from a custom path', async () => {
    const requirer = jest.fn();
    requirer.mockReturnValue(MockConfig.config);
    await readConfig('custom-path', requirer);
    expect(requirer).toHaveBeenCalledWith('custom-path');
  });

  it('requires config files that exports an object', async () => {
    const requirer = jest.fn();
    requirer.mockReturnValue(MockConfig.config[0]);
    const config = await readConfig(undefined, requirer);
    expect(config).toEqual(MockConfig.config);
  });

  it('requires config files that exports an array', async () => {
    const requirer = jest.fn();
    requirer.mockReturnValue(MockConfig.config);
    const config = await readConfig(undefined, requirer);
    expect(config).toEqual(MockConfig.config);
  });

  it('requires config files that exports a function that returns an object', async () => {
    const requirer = jest.fn();
    requirer.mockReturnValue(() => MockConfig.config[0]);
    const config = await readConfig(undefined, requirer);
    expect(config).toEqual(MockConfig.config);
  });

  it('requires config files that exports a function that returns an array', async () => {
    const requirer = jest.fn();
    requirer.mockReturnValue(() => MockConfig.config);
    const config = await readConfig(undefined, requirer);
    expect(config).toEqual(MockConfig.config);
  });
});
