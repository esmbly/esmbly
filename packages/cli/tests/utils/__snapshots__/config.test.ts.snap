// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createConfig creates a new config file 1`] = `
Array [
  Array [
    "root-path/esmbly.config.js",
    "config-module-source",
    Object {
      "overwrite": false,
    },
  ],
]
`;

exports[`createConfig creates a new config file with default values when passing the default option 1`] = `
Array [
  Array [
    "root-path/esmbly.config.js",
    "module.exports = {
  input: [],
  output: [],
  transformers: [],
};
",
    Object {
      "overwrite": false,
    },
  ],
]
`;

exports[`createConfig throws an error if there already is a config file 1`] = `"Config root-path/esmbly.config.js already exists. Remove that file first."`;

exports[`getTemplateConfig resolves to the correct template config 1`] = `Promise {}`;
